rfcs/1440_generic_planner_node.md

```diff
@@ -0,0 +1,238 @@
+  to detect potential problems.
+- It would usually be easier to reason about the security of a static node
+  graph.
+
+## Overview
+
+Moving all node and channel creation to a configuration-based mechanism is a
+significant and potentially disruptive change. This RFC proposes a more gradual
+approach by implementing a re-usable Wasm-based planner node that uses a
+manifest file to do the node and channel creation. The functionality can be
+expanded over time to support more complex scenarios.
+
+The planner node will be responsible for creating nodes, creating channels,
+sending channel handles to the created nodes to ensure that the full node graph
+is created on start-up. It will will also act as a message router, which would
+route HTTP- and GRPC invocations to other nodes based on the associated labels.
+
+## Detailed Design
+
+### Planner Node
+
+The planner node will the the initial node created for an application. The
+manifest file will be passed to the planner node as part of the application's
+`ConfigMap`.
+
+The planner node will create all the nodes and channels as specified with the
+appropriate labels. It will send initialisation messages to each of the nodes,
+providing additional configuration and handles to channels. It is the
+responsibility of each of the nodes to store these channels for future
+communications with other nodes. It will send a filtered version of the config
+map as part of the initialisation message, based on the configuration keys
+specified in the `config_files` field.
+
+### Invocation Routing
+
+The planner node will also be responsible for routing invocations from the gRPC-
+and HTTP server pseudo nodes to the right receiver node based on the associated
+labels. Since it would be possible for multiple nodes to have the same label,
+the planner node would need an indication of which node should be the invocation
+recipient for a specific label.
+
+### Manifest File
+
+The manifest file will be will be passed to the application as part of the
+initial configuration map as a TOML file. The file will be deserialised into the
+following structure:
+
+```rust
+
+struct Node {
+  /// Unique name for the node.
+  name: String,
+
+  /// The label associated with the node.
+  label: Label,
+
+  /// The keys of the files from the config map that must be passed to the node
+  /// as part of the initialisation message. This will be used to generate a
+  /// filtered view of the map that will be sent as part of the initialisation
+  /// message.
+  config_files: Vec<String>,
+
+  /// Whether this node will accept gRPC invocations targeted at the specified
+  /// label.
+  grpc_invocation_receiver: bool,
+
+  /// Whether this node will accept HTTP invocations targeted at the specified
+  /// label.
+  http_invocation_receiver: bool,
+}
+
+struct Channel {
+  /// Unique name of the channel.
+  name: String,
+
+  /// The label associated with the channel.
+  label: Label,
```

Comment:
What if this label depends on the input request?

To my earlier point, if we focus on introducing "label variables" or placeholder, then a subgraph may reuse the same label (which will be bound at runtime) to instantiate various channels or messages within the subgraph.

Reply:
I will update the RFC to cover support for channel and node templates to support dynamic label-based node and channel creation.